File: rust/parity/specs/threat_model_generator.py
"""Parity spec for PiThreatModelGenerator.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}

PARITY CAVEAT: the original Python agent computes
  ``STRIDE_categories = list(set(categories))``
whose ordering is governed by CPython per-process hash randomization
(``PYTHONHASHSEED``). The Python output ordering of ``STRIDE_categories`` is
therefore NON-DETERMINISTIC across runs whenever more than one category is
present, and cannot be byte-matched by any deterministic Rust port. The Rust
port deduplicates preserving first-seen insertion order. Samples that populate
two or more categories may flake on the ``STRIDE_categories`` field. See the
agent's `deviations`.
"""
from _util import load_py_agent

RUST_NAME = "PiThreatModelGenerator"

_mod = load_py_agent("pi_threat_model_generator.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiThreatModelGenerator()
    out = agent.generate_threat_model(_mod.SystemInput(**data))
    return out.model_dump()


SAMPLES = [
    # clean / passing input — no trigger tokens, no categories (deterministic)
    {"input": {"system_desc": "a benign internal batch tool with no external surface"}},
    # database path only -> Information Disclosure + Tampering, risk 60
    {"input": {"system_desc": "service backed by a postgres database and object storage"}},
    # api path only -> Elevation of Privilege + Denial of Service, risk 80
    {"input": {"system_desc": "an api gateway exposing a public endpoint"}},
    # public web path only -> Spoofing, risk 50 (single category, deterministic)
    {"input": {"system_desc": "a public web frontend served to every client"}},
    # all three paths -> 5 threats, max risk 80, multi-category (ORDER MAY FLAKE)
    {"input": {"system_desc": "a system with a database, an api gateway, and a public web frontend"}},
    # short trigger token 'db' substring match -> database path
    {"input": {"system_desc": "uses a small embedded db for caching"}},
    # empty edge input
    {"input": {"system_desc": ""}},
    # non-strict env -> is_secure stays True, status becomes WARN_THREATS
    {"input": {"system_desc": "a public web frontend served to every client"},
     "env": {"PI_SYSTEM_STRICT_MODE": "false"}},
    # strict env explicitly true -> is_secure False
    {"input": {"system_desc": "a public web frontend served to every client"},
     "env": {"PI_SYSTEM_STRICT_MODE": "true"}},
]

# The original agent dedupes via `list(set(categories))` (source line 62), whose
# iteration order is non-deterministic in CPython (per-process hash randomization).
# The Rust port is deterministic; there is no canonical Python order to match, so
# we compare this field as an order-insensitive set.
NORMALIZE = ["STRIDE_categories"]

File: rust/crates/pi-agents/src/agents/threat_model_generator.rs
//! Port of `pi_micro_agents/pi_threat_model_generator.py`.
//!
//! Generates a STRIDE-style threat model from a high-level system description.
//! Behaviour mirrors the Python original line-for-line.
//!
//! PARITY CAVEAT (see parity spec / deviations): the Python original computes
//! `STRIDE_categories = list(set(categories))`. CPython set iteration order for
//! strings is governed by per-process hash randomization (`PYTHONHASHSEED`), so
//! the Python output ordering of `STRIDE_categories` is NON-DETERMINISTIC across
//! runs. This Rust port deduplicates while preserving first-seen insertion
//! order, which is deterministic but will NOT necessarily byte-match Python's
//! scrambled ordering when more than one category is present.

// Note: this agent uses only `.to_lowercase()` and `.contains()`; the Python
// original never calls `.splitlines()`/`.strip()`, so `crate::pyutil` is not needed.
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub system_desc: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub threats: Vec<String>,
    #[serde(rename = "STRIDE_categories")]
    pub stride_categories: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true". Unset => strict (True).
fn is_strict_mode() -> bool {
    match std::env::var("PI_SYSTEM_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn generate_threat_model(input: &Input) -> Output {
    // `desc = input_envelope.system_desc.lower()`
    let desc = input.system_desc.to_lowercase();
    let mut threats: Vec<String> = Vec::new();
    let mut categories: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // Database related threats
    if desc.contains("database") || desc.contains("db") || desc.contains("storage") {
        threats.push(
            "Information Disclosure: Potential compromise of sensitive user databases due to weak access policies."
                .to_string(),
        );
        categories.push("Information Disclosure".to_string());
        threats.push(
            "Tampering: Malicious injection or truncation queries executed directly on storage clusters."
                .to_string(),
        );
        categories.push("Tampering".to_string());
        risk_score = risk_score.max(60.0);
    }

    // API related threats
    if desc.contains("api") || desc.contains("endpoint") || desc.contains("gateway") {
        threats.push(
            "Elevation of Privilege: Unauthenticated attackers abusing broken authorization boundaries."
                .to_string(),
        );
        categories.push("Elevation of Privilege".to_string());
        threats.push(
            "Denial of Service: Volumetric request bursts exhausting thread pools or backend CPU limits."
                .to_string(),
        );
        categories.push("Denial of Service".to_string());
        risk_score = risk_score.max(80.0);
    }

    // Public web interface
    if desc.contains("public web") || desc.contains("frontend") || desc.contains("client") {
        threats.push(
            "Spoofing: Phishing portals imitating production client domain names.".to_string(),
        );
        categories.push("Spoofing".to_string());
        risk_score = risk_score.max(50.0);
    }

    // Ensure categories are unique.
    //
    // Python: `categories = list(set(categories))` — set() dedup with
    // hash-randomized iteration order. We dedup preserving first-seen order.
    // (See module-level PARITY CAVEAT.)
    let categories = dedup_preserve_order(categories);

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec {
        "PASSED".to_string()
    } else {
        "THREATS_IDENTIFIED".to_string()
    };
    if risk_score > 0.0 && is_sec {
        status = "WARN_THREATS".to_string();
    }

    Output {
        is_secure: is_sec,
        threats,
        stride_categories: categories,
        risk_score,
        status,
    }
}

/// Deduplicate a vector of strings keeping the first occurrence's position.
fn dedup_preserve_order(items: Vec<String>) -> Vec<String> {
    let mut seen = std::collections::HashSet::new();
    let mut out = Vec::new();
    for item in items {
        if seen.insert(item.clone()) {
            out.push(item);
        }
    }
    out
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = generate_threat_model(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(desc: &str) -> Output {
        generate_threat_model(&Input {
            system_desc: desc.into(),
        })
    }

    #[test]
    fn clean_input_passes() {
        // Note: "clean app" — but beware substring matches. "clean" does not
        // contain any trigger token; verify a truly benign string.
        let o = run("a benign internal tool");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.threats.is_empty());
        assert!(o.stride_categories.is_empty());
    }

    #[test]
    fn database_flagged_strict() {
        let o = run("backed by a postgres database and storage layer");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 60.0);
        assert_eq!(o.status, "THREATS_IDENTIFIED");
        assert_eq!(o.threats.len(), 2);
        assert_eq!(
            o.stride_categories,
            vec!["Information Disclosure".to_string(), "Tampering".to_string()]
        );
    }

    #[test]
    fn api_takes_highest_risk() {
        let o = run("an api gateway with a database");
        // database -> 60, api -> 80; max == 80
        assert_eq!(o.risk_score, 80.0);
        assert!(!o.is_secure);
    }

    #[test]
    fn empty_input_passes() {
        let o = run("");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
    }
}
